js/packages/mobile-wallet-adapter-protocol/android/src/main/java/com/solanamobile/mobilewalletadapter/reactnative/SolanaMobileWalletAdapterModule.kt

```diff
@@ -0,0 +1,210 @@
+package com.solanamobile.mobilewalletadapter.reactnative
+
+import android.net.Uri
+import android.os.Looper
+import android.util.Log
+import com.facebook.react.bridge.*
+import com.solana.mobilewalletadapter.clientlib.protocol.JsonRpc20Client
+import com.solana.mobilewalletadapter.clientlib.protocol.MobileWalletAdapterClient
+import com.solana.mobilewalletadapter.clientlib.scenario.LocalAssociationScenario
+import com.solana.mobilewalletadapter.clientlib.scenario.Scenario
+import kotlinx.coroutines.*
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.Semaphore
+import org.json.JSONArray
+import org.json.JSONException
+import org.json.JSONObject
+import java.util.concurrent.ExecutionException
+
+class SolanaMobileWalletAdapterModule(reactContext: ReactApplicationContext) :
+    ReactContextBaseJavaModule(reactContext), CoroutineScope {
+
+    override val coroutineContext =
+        Dispatchers.IO + CoroutineName("SolanaMobileWalletAdapterModuleScope") + SupervisorJob()
+
+    companion object {
+        private const val ASSOCIATION_TIMEOUT_MS = 10000
+        private const val CLIENT_TIMEOUT_MS = 90000
+        private var localAssociation: LocalAssociationScenario? = null
+        private var mobileWalletAdapterClient: MobileWalletAdapterClient? = null
+        private var semTerminated: Semaphore? = null
+        private val mutex = Mutex()
+    }
+
+    override fun getName(): String {
+        return "SolanaMobileWalletAdapter"
+    }
+
+    @Throws(JSONException::class)
+    private fun convertMapToJson(readableMap: ReadableMap?): JSONObject {
```

Comment:
Could we directly use an existing library to map it to what we want? (to replace the various methods manipulation JSON in this class)
Like `kotlinx.serialization` or `moshi`?

Reply:
I looked at `kotlinx.serialization` and my eyes glazed over. How on earth `JSONObject` doesn't have a `toMap` method on it is beyond me. Googling ‘react native JSONObject to map’ and watching everyone flail around boggles the mind.